<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8">
    <title>Coverage for src/bam2tensor/functions.py: 75%</title>
    <link rel="icon" sizes="32x32" href="favicon_32_cb_c827f16f.png">
    <link rel="stylesheet" href="style_cb_9ff733b0.css" type="text/css">
    <script src="coverage_html_cb_dd2e7eb5.js" defer></script>
</head>
<body class="pyfile">
<header>
    <div class="content">
        <h1>
            <span class="text">Coverage for </span><b>src&#8201;/&#8201;bam2tensor&#8201;/&#8201;functions.py</b>:
            <span class="pc_cov">75%</span>
        </h1>
        <aside id="help_panel_wrapper">
            <input id="help_panel_state" type="checkbox">
            <label for="help_panel_state">
                <img id="keyboard_icon" src="keybd_closed_cb_900cfef5.png" alt="Show/hide keyboard shortcuts">
            </label>
            <div id="help_panel">
                <p class="legend">Shortcuts on this page</p>
                <div class="keyhelp">
                    <p>
                        <kbd>r</kbd>
                        <kbd>m</kbd>
                        <kbd>x</kbd>
                        <kbd>p</kbd>
                        &nbsp; toggle line displays
                    </p>
                    <p>
                        <kbd>j</kbd>
                        <kbd>k</kbd>
                        &nbsp; next/prev highlighted chunk
                    </p>
                    <p>
                        <kbd>0</kbd> &nbsp; (zero) top of page
                    </p>
                    <p>
                        <kbd>1</kbd> &nbsp; (one) first highlighted chunk
                    </p>
                    <p>
                        <kbd>[</kbd>
                        <kbd>]</kbd>
                        &nbsp; prev/next file
                    </p>
                    <p>
                        <kbd>u</kbd> &nbsp; up to the index
                    </p>
                    <p>
                        <kbd>?</kbd> &nbsp; show/hide this help
                    </p>
                </div>
            </div>
        </aside>
        <h2>
            <span class="text">100 statements &nbsp;</span>
            <button type="button" class="run button_toggle_run" value="run" data-shortcut="r" title="Toggle lines run">80<span class="text"> run</span></button>
            <button type="button" class="mis show_mis button_toggle_mis" value="mis" data-shortcut="m" title="Toggle lines missing">20<span class="text"> missing</span></button>
            <button type="button" class="exc show_exc button_toggle_exc" value="exc" data-shortcut="x" title="Toggle lines excluded">0<span class="text"> excluded</span></button>
            <button type="button" class="par run show_par button_toggle_par" value="par" data-shortcut="p" title="Toggle lines partially run">6<span class="text"> partial</span></button>
        </h2>
        <p class="text">
            <a id="prevFileLink" class="nav" href="z_21b632de40bb3de8_embedding_py.html">&#xab; prev</a> &nbsp; &nbsp;
            <a id="indexLink" class="nav" href="index.html">&Hat; index</a> &nbsp; &nbsp;
            <a id="nextFileLink" class="nav" href="z_a44f0ac069e85531_test_duplication_py.html">&#xbb; next</a>
            &nbsp; &nbsp; &nbsp;
            <a class="nav" href="https://coverage.readthedocs.io/en/7.13.1">coverage.py v7.13.1</a>,
            created at 2026-02-25 22:04 +0000
        </p>
        <aside class="hidden">
            <button type="button" class="button_next_chunk" data-shortcut="j"></button>
            <button type="button" class="button_prev_chunk" data-shortcut="k"></button>
            <button type="button" class="button_top_of_page" data-shortcut="0"></button>
            <button type="button" class="button_first_chunk" data-shortcut="1"></button>
            <button type="button" class="button_prev_file" data-shortcut="["></button>
            <button type="button" class="button_next_file" data-shortcut="]"></button>
            <button type="button" class="button_to_index" data-shortcut="u"></button>
            <button type="button" class="button_show_hide_help" data-shortcut="?"></button>
        </aside>
    </div>
</header>
<main id="source">
    <p class="pln"><span class="n"><a id="t1" href="#t1">1</a></span><span class="t"><span class="str">"""Core methylation extraction functions for bam2tensor.</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t2" href="#t2">2</a></span><span class="t">&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t3" href="#t3">3</a></span><span class="t"><span class="str">This module contains the main function for extracting methylation data from</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t4" href="#t4">4</a></span><span class="t"><span class="str">BAM files: extract_methylation_data_from_bam(). This function reads aligned</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t5" href="#t5">5</a></span><span class="t"><span class="str">bisulfite-sequencing reads and generates a sparse matrix representation of</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t6" href="#t6">6</a></span><span class="t"><span class="str">methylation states at each CpG site.</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t7" href="#t7">7</a></span><span class="t">&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t8" href="#t8">8</a></span><span class="t"><span class="str">The extraction process:</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t9" href="#t9">9</a></span><span class="t"><span class="str">    1. Iterates through each chromosome in the genome embedding</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t10" href="#t10">10</a></span><span class="t"><span class="str">    2. For each aligned read that passes quality filters:</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t11" href="#t11">11</a></span><span class="t"><span class="str">       a. Identifies which CpG sites fall within the read's alignment</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t12" href="#t12">12</a></span><span class="t"><span class="str">       b. Determines the bisulfite strand (forward or reverse)</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t13" href="#t13">13</a></span><span class="t"><span class="str">       c. Extracts methylation state at each covered CpG</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t14" href="#t14">14</a></span><span class="t"><span class="str">    3. Builds a sparse COO matrix with the results</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t15" href="#t15">15</a></span><span class="t">&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t16" href="#t16">16</a></span><span class="t"><span class="str">Methylation State Encoding:</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t17" href="#t17">17</a></span><span class="t"><span class="str">    - 1: Methylated (cytosine remained as C after bisulfite treatment)</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t18" href="#t18">18</a></span><span class="t"><span class="str">    - 0: Unmethylated (cytosine converted to T by bisulfite treatment)</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t19" href="#t19">19</a></span><span class="t"><span class="str">    - -1: No data (indel, SNV, or other non-C/T base at CpG position)</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t20" href="#t20">20</a></span><span class="t">&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t21" href="#t21">21</a></span><span class="t"><span class="str">Supported Aligners:</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t22" href="#t22">22</a></span><span class="t"><span class="str">    The module supports BAM files from aligners that provide strand information:</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t23" href="#t23">23</a></span><span class="t"><span class="str">    - Biscuit: Uses the YD tag (f=forward, r=reverse)</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t24" href="#t24">24</a></span><span class="t"><span class="str">    - gem3/Blueprint: Uses the XB tag (C=forward, G=reverse)</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t25" href="#t25">25</a></span><span class="t">&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t26" href="#t26">26</a></span><span class="t"><span class="str">Example:</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t27" href="#t27">27</a></span><span class="t"><span class="str">    >>> # xdoctest: +SKIP</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t28" href="#t28">28</a></span><span class="t"><span class="str">    >>> from bam2tensor.embedding import GenomeMethylationEmbedding</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t29" href="#t29">29</a></span><span class="t"><span class="str">    >>> from bam2tensor.functions import extract_methylation_data_from_bam</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t30" href="#t30">30</a></span><span class="t"><span class="str">    >>></span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t31" href="#t31">31</a></span><span class="t"><span class="str">    >>> # Create genome embedding</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t32" href="#t32">32</a></span><span class="t"><span class="str">    >>> embedding = GenomeMethylationEmbedding(</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t33" href="#t33">33</a></span><span class="t"><span class="str">    ...     genome_name="hg38",</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t34" href="#t34">34</a></span><span class="t"><span class="str">    ...     expected_chromosomes=["chr1", "chr2"],</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t35" href="#t35">35</a></span><span class="t"><span class="str">    ...     fasta_source="reference.fa",</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t36" href="#t36">36</a></span><span class="t"><span class="str">    ... )</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t37" href="#t37">37</a></span><span class="t"><span class="str">    >>></span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t38" href="#t38">38</a></span><span class="t"><span class="str">    >>> # Extract methylation data</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t39" href="#t39">39</a></span><span class="t"><span class="str">    >>> matrix = extract_methylation_data_from_bam(</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t40" href="#t40">40</a></span><span class="t"><span class="str">    ...     input_bam="sample.bam",</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t41" href="#t41">41</a></span><span class="t"><span class="str">    ...     genome_methylation_embedding=embedding,</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t42" href="#t42">42</a></span><span class="t"><span class="str">    ...     quality_limit=20,</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t43" href="#t43">43</a></span><span class="t"><span class="str">    ... )</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t44" href="#t44">44</a></span><span class="t"><span class="str">    >>></span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t45" href="#t45">45</a></span><span class="t"><span class="str">    >>> print(f"Extracted {matrix.shape[0]} reads, {matrix.nnz} data points")</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t46" href="#t46">46</a></span><span class="t"><span class="str">"""</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t47" href="#t47">47</a></span><span class="t">&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t48" href="#t48">48</a></span><span class="t"><span class="key">import</span> <span class="nam">scipy</span><span class="op">.</span><span class="nam">sparse</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t49" href="#t49">49</a></span><span class="t"><span class="key">import</span> <span class="nam">pysam</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t50" href="#t50">50</a></span><span class="t"><span class="key">import</span> <span class="nam">bisect</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t51" href="#t51">51</a></span><span class="t">&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t52" href="#t52">52</a></span><span class="t"><span class="key">from</span> <span class="nam">tqdm</span> <span class="key">import</span> <span class="nam">tqdm</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t53" href="#t53">53</a></span><span class="t"><span class="key">from</span> <span class="nam">bam2tensor</span><span class="op">.</span><span class="nam">embedding</span> <span class="key">import</span> <span class="nam">GenomeMethylationEmbedding</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t54" href="#t54">54</a></span><span class="t">&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t55" href="#t55">55</a></span><span class="t">&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t56" href="#t56">56</a></span><span class="t"><span class="key">def</span> <span class="nam">extract_methylation_data_from_bam</span><span class="op">(</span>&nbsp;</span><span class="r"></span></p>
    <p class="run run2"><span class="n"><a id="t57" href="#t57">57</a></span><span class="t">    <span class="nam">input_bam</span><span class="op">:</span> <span class="nam">str</span><span class="op">,</span>&nbsp;</span><span class="r"></span></p>
    <p class="run run2"><span class="n"><a id="t58" href="#t58">58</a></span><span class="t">    <span class="nam">genome_methylation_embedding</span><span class="op">:</span> <span class="nam">GenomeMethylationEmbedding</span><span class="op">,</span>&nbsp;</span><span class="r"></span></p>
    <p class="run run2"><span class="n"><a id="t59" href="#t59">59</a></span><span class="t">    <span class="nam">quality_limit</span><span class="op">:</span> <span class="nam">int</span> <span class="op">=</span> <span class="num">20</span><span class="op">,</span>&nbsp;</span><span class="r"></span></p>
    <p class="run run2"><span class="n"><a id="t60" href="#t60">60</a></span><span class="t">    <span class="nam">verbose</span><span class="op">:</span> <span class="nam">bool</span> <span class="op">=</span> <span class="key">False</span><span class="op">,</span>&nbsp;</span><span class="r"></span></p>
    <p class="run run2"><span class="n"><a id="t61" href="#t61">61</a></span><span class="t">    <span class="nam">debug</span><span class="op">:</span> <span class="nam">bool</span> <span class="op">=</span> <span class="key">False</span><span class="op">,</span>&nbsp;</span><span class="r"></span></p>
    <p class="run run2"><span class="n"><a id="t62" href="#t62">62</a></span><span class="t"><span class="op">)</span> <span class="op">-></span> <span class="nam">scipy</span><span class="op">.</span><span class="nam">sparse</span><span class="op">.</span><span class="nam">coo_matrix</span><span class="op">:</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t63" href="#t63">63</a></span><span class="t">    <span class="str">"""Extract read-level CpG methylation data from a BAM file.</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t64" href="#t64">64</a></span><span class="t">&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t65" href="#t65">65</a></span><span class="t"><span class="str">    Parses a bisulfite-sequencing BAM file and extracts methylation states</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t66" href="#t66">66</a></span><span class="t"><span class="str">    at each CpG site covered by aligned reads. Returns a sparse matrix where</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t67" href="#t67">67</a></span><span class="t"><span class="str">    each row represents a unique read and each column represents a CpG site.</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t68" href="#t68">68</a></span><span class="t">&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t69" href="#t69">69</a></span><span class="t"><span class="str">    The function applies several filters to reads:</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t70" href="#t70">70</a></span><span class="t"><span class="str">        - Mapping quality must be >= quality_limit (default 20)</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t71" href="#t71">71</a></span><span class="t"><span class="str">        - Duplicate reads are excluded</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t72" href="#t72">72</a></span><span class="t"><span class="str">        - QC-failed reads are excluded</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t73" href="#t73">73</a></span><span class="t"><span class="str">        - Secondary and supplementary alignments are excluded</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t74" href="#t74">74</a></span><span class="t"><span class="str">        - Only reads on the bisulfite-converted parent strand are processed</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t75" href="#t75">75</a></span><span class="t">&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t76" href="#t76">76</a></span><span class="t"><span class="str">    For each qualifying read, the function:</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t77" href="#t77">77</a></span><span class="t"><span class="str">        1. Uses binary search to find CpG sites within the read's alignment</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t78" href="#t78">78</a></span><span class="t"><span class="str">        2. Checks the bisulfite strand tag (YD or XB) to determine strand</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t79" href="#t79">79</a></span><span class="t"><span class="str">        3. Extracts the base at each CpG position from the read sequence</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t80" href="#t80">80</a></span><span class="t"><span class="str">        4. Records methylation state: C=methylated(1), T=unmethylated(0), other=-1</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t81" href="#t81">81</a></span><span class="t">&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t82" href="#t82">82</a></span><span class="t"><span class="str">    Args:</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t83" href="#t83">83</a></span><span class="t"><span class="str">        input_bam: Path to the input BAM file. The file must be indexed</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t84" href="#t84">84</a></span><span class="t"><span class="str">            (have a corresponding .bam.bai file).</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t85" href="#t85">85</a></span><span class="t"><span class="str">        genome_methylation_embedding: A GenomeMethylationEmbedding object</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t86" href="#t86">86</a></span><span class="t"><span class="str">            that defines the CpG site positions and provides coordinate</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t87" href="#t87">87</a></span><span class="t"><span class="str">            conversion methods.</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t88" href="#t88">88</a></span><span class="t"><span class="str">        quality_limit: Minimum mapping quality (MAPQ) threshold for reads.</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t89" href="#t89">89</a></span><span class="t"><span class="str">            Reads with MAPQ below this value are skipped. Default is 20,</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t90" href="#t90">90</a></span><span class="t"><span class="str">            which excludes reads mapping to multiple locations equally well.</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t91" href="#t91">91</a></span><span class="t"><span class="str">        verbose: If True, display a progress bar and print the total read</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t92" href="#t92">92</a></span><span class="t"><span class="str">            count. Useful for monitoring progress on large files.</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t93" href="#t93">93</a></span><span class="t"><span class="str">        debug: If True, enable extensive validation and debug output.</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t94" href="#t94">94</a></span><span class="t"><span class="str">            Prints per-read information and validates that each read is</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t95" href="#t95">95</a></span><span class="t"><span class="str">            only processed once. Significantly slower.</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t96" href="#t96">96</a></span><span class="t">&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t97" href="#t97">97</a></span><span class="t"><span class="str">    Returns:</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t98" href="#t98">98</a></span><span class="t"><span class="str">        A scipy.sparse.coo_matrix with shape (n_reads, n_cpg_sites) where:</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t99" href="#t99">99</a></span><span class="t"><span class="str">            - n_reads is the number of reads that passed filters and covered</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t100" href="#t100">100</a></span><span class="t"><span class="str">              at least one CpG site</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t101" href="#t101">101</a></span><span class="t"><span class="str">            - n_cpg_sites is genome_methylation_embedding.total_cpg_sites</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t102" href="#t102">102</a></span><span class="t"><span class="str">            - Values are: 1 (methylated), 0 (unmethylated), -1 (no data)</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t103" href="#t103">103</a></span><span class="t">&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t104" href="#t104">104</a></span><span class="t"><span class="str">        The matrix uses COO format for efficient construction. Convert to</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t105" href="#t105">105</a></span><span class="t"><span class="str">        CSR (tocsr()) for row slicing or CSC (tocsc()) for column slicing.</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t106" href="#t106">106</a></span><span class="t">&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t107" href="#t107">107</a></span><span class="t"><span class="str">    Raises:</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t108" href="#t108">108</a></span><span class="t"><span class="str">        FileNotFoundError: If the BAM file index (.bam.bai) is missing.</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t109" href="#t109">109</a></span><span class="t"><span class="str">            The BAM file must be indexed with `samtools index`.</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t110" href="#t110">110</a></span><span class="t">&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t111" href="#t111">111</a></span><span class="t"><span class="str">    Example:</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t112" href="#t112">112</a></span><span class="t"><span class="str">        >>> # xdoctest: +SKIP</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t113" href="#t113">113</a></span><span class="t"><span class="str">        >>> from bam2tensor.embedding import GenomeMethylationEmbedding</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t114" href="#t114">114</a></span><span class="t"><span class="str">        >>> from bam2tensor.functions import extract_methylation_data_from_bam</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t115" href="#t115">115</a></span><span class="t"><span class="str">        >>> import scipy.sparse</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t116" href="#t116">116</a></span><span class="t"><span class="str">        >>></span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t117" href="#t117">117</a></span><span class="t"><span class="str">        >>> # Create embedding (or load from cache)</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t118" href="#t118">118</a></span><span class="t"><span class="str">        >>> embedding = GenomeMethylationEmbedding(</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t119" href="#t119">119</a></span><span class="t"><span class="str">        ...     genome_name="hg38",</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t120" href="#t120">120</a></span><span class="t"><span class="str">        ...     expected_chromosomes=["chr1", "chr22"],</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t121" href="#t121">121</a></span><span class="t"><span class="str">        ...     fasta_source="GRCh38.fa",</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t122" href="#t122">122</a></span><span class="t"><span class="str">        ... )</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t123" href="#t123">123</a></span><span class="t"><span class="str">        >>></span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t124" href="#t124">124</a></span><span class="t"><span class="str">        >>> # Extract methylation data</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t125" href="#t125">125</a></span><span class="t"><span class="str">        >>> matrix = extract_methylation_data_from_bam(</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t126" href="#t126">126</a></span><span class="t"><span class="str">        ...     input_bam="sample.bam",</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t127" href="#t127">127</a></span><span class="t"><span class="str">        ...     genome_methylation_embedding=embedding,</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t128" href="#t128">128</a></span><span class="t"><span class="str">        ...     quality_limit=30,  # Stricter quality filter</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t129" href="#t129">129</a></span><span class="t"><span class="str">        ...     verbose=True,</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t130" href="#t130">130</a></span><span class="t"><span class="str">        ... )</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t131" href="#t131">131</a></span><span class="t"><span class="str">        >>></span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t132" href="#t132">132</a></span><span class="t"><span class="str">        >>> # Analyze results</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t133" href="#t133">133</a></span><span class="t"><span class="str">        >>> print(f"Reads with CpG data: {matrix.shape[0]:,}")</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t134" href="#t134">134</a></span><span class="t"><span class="str">        >>> print(f"Total CpG sites: {matrix.shape[1]:,}")</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t135" href="#t135">135</a></span><span class="t"><span class="str">        >>> print(f"Data points: {matrix.nnz:,}")</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t136" href="#t136">136</a></span><span class="t"><span class="str">        >>></span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t137" href="#t137">137</a></span><span class="t"><span class="str">        >>> # Save to file</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t138" href="#t138">138</a></span><span class="t"><span class="str">        >>> scipy.sparse.save_npz("sample.methylation.npz", matrix)</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t139" href="#t139">139</a></span><span class="t">&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t140" href="#t140">140</a></span><span class="t"><span class="str">    Note:</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t141" href="#t141">141</a></span><span class="t"><span class="str">        The function processes chromosomes in the order they appear in</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t142" href="#t142">142</a></span><span class="t"><span class="str">        genome_methylation_embedding.cpg_sites_dict. Reads are numbered</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t143" href="#t143">143</a></span><span class="t"><span class="str">        sequentially as they are encountered across all chromosomes.</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t144" href="#t144">144</a></span><span class="t">&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t145" href="#t145">145</a></span><span class="t"><span class="str">    See Also:</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t146" href="#t146">146</a></span><span class="t"><span class="str">        GenomeMethylationEmbedding: For creating the genome embedding.</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t147" href="#t147">147</a></span><span class="t"><span class="str">        scipy.sparse.coo_matrix: Documentation on the output format.</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t148" href="#t148">148</a></span><span class="t"><span class="str">    """</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t149" href="#t149">149</a></span><span class="t">    <span class="key">try</span><span class="op">:</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t150" href="#t150">150</a></span><span class="t">        <span class="nam">input_bam_object</span> <span class="op">=</span> <span class="nam">pysam</span><span class="op">.</span><span class="nam">AlignmentFile</span><span class="op">(</span>  <span class="com"># type: ignore # pylint: disable=no-member</span>&nbsp;</span><span class="r"></span></p>
    <p class="run run2"><span class="n"><a id="t151" href="#t151">151</a></span><span class="t">            <span class="nam">input_bam</span><span class="op">,</span> <span class="str">"rb"</span><span class="op">,</span> <span class="nam">require_index</span><span class="op">=</span><span class="key">True</span><span class="op">,</span> <span class="nam">threads</span><span class="op">=</span><span class="num">1</span>&nbsp;</span><span class="r"></span></p>
    <p class="run run2"><span class="n"><a id="t152" href="#t152">152</a></span><span class="t">        <span class="op">)</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t153" href="#t153">153</a></span><span class="t">    <span class="key">except</span> <span class="nam">FileNotFoundError</span> <span class="key">as</span> <span class="nam">exc</span><span class="op">:</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t154" href="#t154">154</a></span><span class="t">        <span class="key">raise</span> <span class="nam">FileNotFoundError</span><span class="op">(</span><span class="fst">f"</span><span class="fst">Index missing for bam file?: </span><span class="op">{</span><span class="nam">input_bam</span><span class="op">}</span><span class="fst">"</span><span class="op">)</span> <span class="key">from</span> <span class="nam">exc</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t155" href="#t155">155</a></span><span class="t">&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t156" href="#t156">156</a></span><span class="t">    <span class="key">if</span> <span class="nam">verbose</span><span class="op">:</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t157" href="#t157">157</a></span><span class="t">        <span class="nam">print</span><span class="op">(</span><span class="fst">f"</span><span class="fst">\tTotal reads: </span><span class="op">{</span><span class="nam">input_bam_object</span><span class="op">.</span><span class="nam">mapped</span><span class="op">:</span><span class="fst">,</span><span class="op">}</span><span class="fst">\n</span><span class="fst">"</span><span class="op">)</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t158" href="#t158">158</a></span><span class="t">    <span class="key">if</span> <span class="nam">debug</span><span class="op">:</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t159" href="#t159">159</a></span><span class="t">        <span class="nam">debug_read_name_to_row_number</span> <span class="op">=</span> <span class="op">{</span><span class="op">}</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t160" href="#t160">160</a></span><span class="t">&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t161" href="#t161">161</a></span><span class="t">    <span class="com"># Temporary storage for loading into a COO matrix</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t162" href="#t162">162</a></span><span class="t">    <span class="com"># For a COO, we want three lists:</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t163" href="#t163">163</a></span><span class="t">    <span class="com"># row: the read number (we'll store a read name -> ID dict perhaps?)</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t164" href="#t164">164</a></span><span class="t">    <span class="com"># column: cpg #</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t165" href="#t165">165</a></span><span class="t">    <span class="com"># data: methylation state (1 [methylated], 0 [unmethylated], and -1 [no data/snv/indel])</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t166" href="#t166">166</a></span><span class="t">    <span class="nam">read_number</span> <span class="op">=</span> <span class="num">0</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t167" href="#t167">167</a></span><span class="t">&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t168" href="#t168">168</a></span><span class="t">    <span class="nam">coo_row</span> <span class="op">=</span> <span class="op">[</span><span class="op">]</span>  <span class="com"># Read number</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t169" href="#t169">169</a></span><span class="t">    <span class="nam">coo_col</span> <span class="op">=</span> <span class="op">[</span><span class="op">]</span>  <span class="com"># CpG number (embedding)</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t170" href="#t170">170</a></span><span class="t">    <span class="nam">coo_data</span> <span class="op">=</span> <span class="op">[</span><span class="op">]</span>  <span class="com"># Methylation state</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t171" href="#t171">171</a></span><span class="t">&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t172" href="#t172">172</a></span><span class="t">    <span class="com"># This is slow, but we only run it once and store the results for later</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t173" href="#t173">173</a></span><span class="t">    <span class="key">for</span> <span class="nam">chrom</span><span class="op">,</span> <span class="nam">cpg_sites</span> <span class="key">in</span> <span class="nam">tqdm</span><span class="op">(</span>&nbsp;</span><span class="r"></span></p>
    <p class="run run2"><span class="n"><a id="t174" href="#t174">174</a></span><span class="t">        <span class="nam">genome_methylation_embedding</span><span class="op">.</span><span class="nam">cpg_sites_dict</span><span class="op">.</span><span class="nam">items</span><span class="op">(</span><span class="op">)</span><span class="op">,</span>&nbsp;</span><span class="r"></span></p>
    <p class="run run2"><span class="n"><a id="t175" href="#t175">175</a></span><span class="t">        <span class="nam">disable</span><span class="op">=</span><span class="key">not</span> <span class="nam">verbose</span><span class="op">,</span>&nbsp;</span><span class="r"></span></p>
    <p class="run run2"><span class="n"><a id="t176" href="#t176">176</a></span><span class="t">    <span class="op">)</span><span class="op">:</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t177" href="#t177">177</a></span><span class="t">        <span class="com"># Ensure this chromosome exists in the BAM file</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t178" href="#t178">178</a></span><span class="t">        <span class="key">try</span><span class="op">:</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t179" href="#t179">179</a></span><span class="t">            <span class="nam">iter_reads</span> <span class="op">=</span> <span class="nam">input_bam_object</span><span class="op">.</span><span class="nam">fetch</span><span class="op">(</span><span class="nam">contig</span><span class="op">=</span><span class="nam">chrom</span><span class="op">)</span>&nbsp;</span><span class="r"></span></p>
    <p class="mis show_mis"><span class="n"><a id="t180" href="#t180">180</a></span><span class="t">        <span class="key">except</span> <span class="nam">ValueError</span><span class="op">:</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t181" href="#t181">181</a></span><span class="t">            <span class="com"># Chromosome not in BAM</span>&nbsp;</span><span class="r"></span></p>
    <p class="mis show_mis"><span class="n"><a id="t182" href="#t182">182</a></span><span class="t">            <span class="key">if</span> <span class="nam">verbose</span><span class="op">:</span>&nbsp;</span><span class="r"></span></p>
    <p class="mis show_mis"><span class="n"><a id="t183" href="#t183">183</a></span><span class="t">                <span class="nam">tqdm</span><span class="op">.</span><span class="nam">write</span><span class="op">(</span><span class="fst">f"</span><span class="fst">Chromosome </span><span class="op">{</span><span class="nam">chrom</span><span class="op">}</span><span class="fst"> not found in BAM file, skipping.</span><span class="fst">"</span><span class="op">)</span>&nbsp;</span><span class="r"></span></p>
    <p class="mis show_mis"><span class="n"><a id="t184" href="#t184">184</a></span><span class="t">            <span class="key">continue</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t185" href="#t185">185</a></span><span class="t">&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t186" href="#t186">186</a></span><span class="t">        <span class="key">for</span> <span class="nam">aligned_segment</span> <span class="key">in</span> <span class="nam">iter_reads</span><span class="op">:</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t187" href="#t187">187</a></span><span class="t">            <span class="key">if</span> <span class="nam">aligned_segment</span><span class="op">.</span><span class="nam">mapping_quality</span> <span class="op">&lt;</span> <span class="nam">quality_limit</span><span class="op">:</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t188" href="#t188">188</a></span><span class="t">                <span class="key">continue</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t189" href="#t189">189</a></span><span class="t">            <span class="key">if</span> <span class="nam">aligned_segment</span><span class="op">.</span><span class="nam">is_duplicate</span><span class="op">:</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t190" href="#t190">190</a></span><span class="t">                <span class="key">continue</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t191" href="#t191">191</a></span><span class="t">            <span class="key">if</span> <span class="nam">aligned_segment</span><span class="op">.</span><span class="nam">is_qcfail</span><span class="op">:</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t192" href="#t192">192</a></span><span class="t">                <span class="key">continue</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t193" href="#t193">193</a></span><span class="t">            <span class="key">if</span> <span class="nam">aligned_segment</span><span class="op">.</span><span class="nam">is_secondary</span><span class="op">:</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t194" href="#t194">194</a></span><span class="t">                <span class="key">continue</span>&nbsp;</span><span class="r"></span></p>
    <p class="par run show_par"><span class="n"><a id="t195" href="#t195">195</a></span><span class="t">            <span class="key">if</span> <span class="nam">aligned_segment</span><span class="op">.</span><span class="nam">is_supplementary</span><span class="op">:</span>&nbsp;</span><span class="r"><span class="annotate short">195&#x202F;&#x219B;&#x202F;196</span><span class="annotate long">line 195 didn't jump to line 196 because the condition on line 195 was never true</span></span></p>
    <p class="mis show_mis"><span class="n"><a id="t196" href="#t196">196</a></span><span class="t">                <span class="key">continue</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t197" href="#t197">197</a></span><span class="t">&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t198" href="#t198">198</a></span><span class="t">            <span class="com"># Use bisect to find CpGs covered by this read</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t199" href="#t199">199</a></span><span class="t">            <span class="com"># aligned_segment.reference_start is 0-based inclusive</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t200" href="#t200">200</a></span><span class="t">            <span class="com"># aligned_segment.reference_end is 0-based exclusive</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t201" href="#t201">201</a></span><span class="t">            <span class="com"># cpg_sites are 1-based</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t202" href="#t202">202</a></span><span class="t">&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t203" href="#t203">203</a></span><span class="t">            <span class="com"># We want CpGs where: read_start &lt; cpg_pos &lt;= read_end (in 1-based terms: read_start+1 &lt;= cpg &lt;= read_end)</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t204" href="#t204">204</a></span><span class="t">            <span class="nam">start_idx</span> <span class="op">=</span> <span class="nam">bisect</span><span class="op">.</span><span class="nam">bisect_left</span><span class="op">(</span>&nbsp;</span><span class="r"></span></p>
    <p class="run run2"><span class="n"><a id="t205" href="#t205">205</a></span><span class="t">                <span class="nam">cpg_sites</span><span class="op">,</span> <span class="nam">aligned_segment</span><span class="op">.</span><span class="nam">reference_start</span> <span class="op">+</span> <span class="num">1</span>&nbsp;</span><span class="r"></span></p>
    <p class="run run2"><span class="n"><a id="t206" href="#t206">206</a></span><span class="t">            <span class="op">)</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t207" href="#t207">207</a></span><span class="t">            <span class="nam">end_idx</span> <span class="op">=</span> <span class="nam">bisect</span><span class="op">.</span><span class="nam">bisect_right</span><span class="op">(</span><span class="nam">cpg_sites</span><span class="op">,</span> <span class="nam">aligned_segment</span><span class="op">.</span><span class="nam">reference_end</span><span class="op">)</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t208" href="#t208">208</a></span><span class="t">&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t209" href="#t209">209</a></span><span class="t">            <span class="com"># If no CpGs in this read, skip</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t210" href="#t210">210</a></span><span class="t">            <span class="key">if</span> <span class="nam">start_idx</span> <span class="op">>=</span> <span class="nam">end_idx</span><span class="op">:</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t211" href="#t211">211</a></span><span class="t">                <span class="key">continue</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t212" href="#t212">212</a></span><span class="t">&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t213" href="#t213">213</a></span><span class="t">            <span class="nam">cpgs_within_read</span> <span class="op">=</span> <span class="nam">cpg_sites</span><span class="op">[</span><span class="nam">start_idx</span><span class="op">:</span><span class="nam">end_idx</span><span class="op">]</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t214" href="#t214">214</a></span><span class="t">            <span class="nam">cpgs_within_read_set</span> <span class="op">=</span> <span class="nam">set</span><span class="op">(</span><span class="nam">cpgs_within_read</span><span class="op">)</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t215" href="#t215">215</a></span><span class="t">&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t216" href="#t216">216</a></span><span class="t">            <span class="key">if</span> <span class="nam">debug</span><span class="op">:</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t217" href="#t217">217</a></span><span class="t">                <span class="nam">print</span><span class="op">(</span><span class="fst">f"</span><span class="fst">Query: </span><span class="op">{</span><span class="nam">aligned_segment</span><span class="op">.</span><span class="nam">query_name</span><span class="op">}</span><span class="fst">"</span><span class="op">)</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t218" href="#t218">218</a></span><span class="t">                <span class="com"># Validity tests</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t219" href="#t219">219</a></span><span class="t">                <span class="key">assert</span> <span class="key">not</span> <span class="nam">aligned_segment</span><span class="op">.</span><span class="nam">is_unmapped</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t220" href="#t220">220</a></span><span class="t">                <span class="key">assert</span> <span class="nam">aligned_segment</span><span class="op">.</span><span class="nam">is_supplementary</span> <span class="key">is</span> <span class="key">False</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t221" href="#t221">221</a></span><span class="t">&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t222" href="#t222">222</a></span><span class="t">                <span class="com"># Ensure alignment methylation tags exist</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t223" href="#t223">223</a></span><span class="t">                <span class="key">assert</span> <span class="nam">aligned_segment</span><span class="op">.</span><span class="nam">has_tag</span><span class="op">(</span>&nbsp;</span><span class="r"></span></p>
    <p class="run run2"><span class="n"><a id="t224" href="#t224">224</a></span><span class="t">                    <span class="str">"MD"</span>&nbsp;</span><span class="r"></span></p>
    <p class="run run2"><span class="n"><a id="t225" href="#t225">225</a></span><span class="t">                <span class="op">)</span>  <span class="com"># Location of mismatches (methylation)</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t226" href="#t226">226</a></span><span class="t">                <span class="key">assert</span> <span class="nam">aligned_segment</span><span class="op">.</span><span class="nam">has_tag</span><span class="op">(</span>&nbsp;</span><span class="r"></span></p>
    <p class="run run2"><span class="n"><a id="t227" href="#t227">227</a></span><span class="t">                    <span class="str">"YD"</span>&nbsp;</span><span class="r"></span></p>
    <p class="run run2"><span class="n"><a id="t228" href="#t228">228</a></span><span class="t">                <span class="op">)</span>  <span class="com"># Bisulfite conversion strand label (f: OT/CTOT C->T or r: OB/CTOB G->A)</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t229" href="#t229">229</a></span><span class="t">&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t230" href="#t230">230</a></span><span class="t">                <span class="com"># Ensure each read is only seen once</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t231" href="#t231">231</a></span><span class="t">                <span class="key">assert</span> <span class="op">(</span>&nbsp;</span><span class="r"></span></p>
    <p class="run run2"><span class="n"><a id="t232" href="#t232">232</a></span><span class="t">                    <span class="nam">aligned_segment</span><span class="op">.</span><span class="nam">query_name</span> <span class="key">not</span> <span class="key">in</span> <span class="nam">debug_read_name_to_row_number</span>&nbsp;</span><span class="r"></span></p>
    <p class="run run2"><span class="n"><a id="t233" href="#t233">233</a></span><span class="t">                <span class="op">)</span><span class="op">,</span> <span class="str">"Read seen twice!"</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t234" href="#t234">234</a></span><span class="t">                <span class="nam">debug_read_name_to_row_number</span><span class="op">[</span>&nbsp;</span><span class="r"></span></p>
    <p class="run run2"><span class="n"><a id="t235" href="#t235">235</a></span><span class="t">                    <span class="nam">aligned_segment</span><span class="op">.</span><span class="nam">query_name</span>  <span class="com"># type: ignore</span>&nbsp;</span><span class="r"></span></p>
    <p class="run run2"><span class="n"><a id="t236" href="#t236">236</a></span><span class="t">                    <span class="op">+</span> <span class="op">(</span><span class="str">"_1"</span> <span class="key">if</span> <span class="nam">aligned_segment</span><span class="op">.</span><span class="nam">is_read1</span> <span class="key">else</span> <span class="str">"_2"</span><span class="op">)</span>&nbsp;</span><span class="r"></span></p>
    <p class="run run2"><span class="n"><a id="t237" href="#t237">237</a></span><span class="t">                <span class="op">]</span> <span class="op">=</span> <span class="nam">read_number</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t238" href="#t238">238</a></span><span class="t">&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t239" href="#t239">239</a></span><span class="t">            <span class="com"># TODO: We ignore paired/unpaired read status for now. Should we treat paired reads / overlapping reads differently?</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t240" href="#t240">240</a></span><span class="t">&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t241" href="#t241">241</a></span><span class="t">            <span class="com">## Read tags and ensure we're on the correct bisulfite-converted strand</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t242" href="#t242">242</a></span><span class="t">            <span class="nam">bisulfite_parent_strand_is_reverse</span> <span class="op">=</span> <span class="key">None</span>&nbsp;</span><span class="r"></span></p>
    <p class="par run show_par"><span class="n"><a id="t243" href="#t243">243</a></span><span class="t">            <span class="key">if</span> <span class="nam">aligned_segment</span><span class="op">.</span><span class="nam">has_tag</span><span class="op">(</span><span class="str">"YD"</span><span class="op">)</span><span class="op">:</span>  <span class="com"># Biscuit tag</span>&nbsp;</span><span class="r"><span class="annotate short">243&#x202F;&#x219B;&#x202F;251</span><span class="annotate long">line 243 didn't jump to line 251 because the condition on line 243 was always true</span></span></p>
    <p class="run"><span class="n"><a id="t244" href="#t244">244</a></span><span class="t">                <span class="nam">yd_tag</span> <span class="op">=</span> <span class="nam">aligned_segment</span><span class="op">.</span><span class="nam">get_tag</span><span class="op">(</span><span class="str">"YD"</span><span class="op">)</span>&nbsp;</span><span class="r"></span></p>
    <p class="par run show_par"><span class="n"><a id="t245" href="#t245">245</a></span><span class="t">                <span class="key">if</span> <span class="nam">yd_tag</span> <span class="op">==</span> <span class="str">"f"</span><span class="op">:</span>  <span class="com"># Forward = C&#8594;T</span>&nbsp;</span><span class="r"><span class="annotate short">245&#x202F;&#x219B;&#x202F;248</span><span class="annotate long">line 245 didn't jump to line 248 because the condition on line 245 was always true</span></span></p>
    <p class="pln"><span class="n"><a id="t246" href="#t246">246</a></span><span class="t">                    <span class="com"># This read derives from OT/CTOT strand: C->T substitutions matter (C = methylated, T = unmethylated),</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t247" href="#t247">247</a></span><span class="t">                    <span class="nam">bisulfite_parent_strand_is_reverse</span> <span class="op">=</span> <span class="key">False</span>&nbsp;</span><span class="r"></span></p>
    <p class="mis show_mis"><span class="n"><a id="t248" href="#t248">248</a></span><span class="t">                <span class="key">elif</span> <span class="nam">yd_tag</span> <span class="op">==</span> <span class="str">"r"</span><span class="op">:</span>  <span class="com"># Reverse = G&#8594;A</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t249" href="#t249">249</a></span><span class="t">                    <span class="com"># This read derives from the OB/CTOB strand: G->A substitutions matter (G = methylated, A = unmethylated)</span>&nbsp;</span><span class="r"></span></p>
    <p class="mis show_mis"><span class="n"><a id="t250" href="#t250">250</a></span><span class="t">                    <span class="nam">bisulfite_parent_strand_is_reverse</span> <span class="op">=</span> <span class="key">True</span>&nbsp;</span><span class="r"></span></p>
    <p class="mis show_mis"><span class="n"><a id="t251" href="#t251">251</a></span><span class="t">            <span class="key">elif</span> <span class="nam">aligned_segment</span><span class="op">.</span><span class="nam">has_tag</span><span class="op">(</span><span class="str">"XB"</span><span class="op">)</span><span class="op">:</span>  <span class="com"># gem3 / blueprint tag</span>&nbsp;</span><span class="r"></span></p>
    <p class="mis show_mis"><span class="n"><a id="t252" href="#t252">252</a></span><span class="t">                <span class="nam">xb_tag</span> <span class="op">=</span> <span class="nam">aligned_segment</span><span class="op">.</span><span class="nam">get_tag</span><span class="op">(</span>&nbsp;</span><span class="r"></span></p>
    <p class="mis mis2 show_mis"><span class="n"><a id="t253" href="#t253">253</a></span><span class="t">                    <span class="str">"XB"</span>&nbsp;</span><span class="r"></span></p>
    <p class="mis mis2 show_mis"><span class="n"><a id="t254" href="#t254">254</a></span><span class="t">                <span class="op">)</span>  <span class="com"># XB:C = Forward / Reference was CG</span>&nbsp;</span><span class="r"></span></p>
    <p class="mis show_mis"><span class="n"><a id="t255" href="#t255">255</a></span><span class="t">                <span class="key">if</span> <span class="nam">xb_tag</span> <span class="op">==</span> <span class="str">"C"</span><span class="op">:</span>&nbsp;</span><span class="r"></span></p>
    <p class="mis show_mis"><span class="n"><a id="t256" href="#t256">256</a></span><span class="t">                    <span class="nam">bisulfite_parent_strand_is_reverse</span> <span class="op">=</span> <span class="key">False</span>&nbsp;</span><span class="r"></span></p>
    <p class="mis show_mis"><span class="n"><a id="t257" href="#t257">257</a></span><span class="t">                <span class="key">elif</span> <span class="nam">xb_tag</span> <span class="op">==</span> <span class="str">"G"</span><span class="op">:</span>  <span class="com"># XB:G = Reverse / Reference was GA</span>&nbsp;</span><span class="r"></span></p>
    <p class="mis show_mis"><span class="n"><a id="t258" href="#t258">258</a></span><span class="t">                    <span class="nam">bisulfite_parent_strand_is_reverse</span> <span class="op">=</span> <span class="key">True</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t259" href="#t259">259</a></span><span class="t">&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t260" href="#t260">260</a></span><span class="t">            <span class="com"># We have paired-end reads; one half should (the "parent strand") has the methylation data.</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t261" href="#t261">261</a></span><span class="t">            <span class="com"># The other half (the "daughter strand") was the complement created by PCR, which we don't care about.</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t262" href="#t262">262</a></span><span class="t">            <span class="key">if</span> <span class="nam">bisulfite_parent_strand_is_reverse</span> <span class="op">!=</span> <span class="nam">aligned_segment</span><span class="op">.</span><span class="nam">is_reverse</span><span class="op">:</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t263" href="#t263">263</a></span><span class="t">                <span class="com"># Skip if we're not on the bisulfite-converted parent strand.</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t264" href="#t264">264</a></span><span class="t">                <span class="key">if</span> <span class="nam">debug</span><span class="op">:</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t265" href="#t265">265</a></span><span class="t">                    <span class="nam">print</span><span class="op">(</span><span class="str">"\tNot on methylated strand, ignoring."</span><span class="op">)</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t266" href="#t266">266</a></span><span class="t">                <span class="key">continue</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t267" href="#t267">267</a></span><span class="t">&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t268" href="#t268">268</a></span><span class="t">            <span class="com"># get_aligned_pairs returns a list of tuples of (read_pos, ref_pos)</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t269" href="#t269">269</a></span><span class="t">            <span class="com"># We filter this to only include the specific CpG sites from above</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t270" href="#t270">270</a></span><span class="t">            <span class="nam">this_segment_cpgs</span> <span class="op">=</span> <span class="op">[</span>&nbsp;</span><span class="r"></span></p>
    <p class="run run2"><span class="n"><a id="t271" href="#t271">271</a></span><span class="t">                <span class="nam">e</span>&nbsp;</span><span class="r"></span></p>
    <p class="run run2"><span class="n"><a id="t272" href="#t272">272</a></span><span class="t">                <span class="key">for</span> <span class="nam">e</span> <span class="key">in</span> <span class="nam">aligned_segment</span><span class="op">.</span><span class="nam">get_aligned_pairs</span><span class="op">(</span><span class="nam">matches_only</span><span class="op">=</span><span class="key">True</span><span class="op">)</span>&nbsp;</span><span class="r"></span></p>
    <p class="run run2"><span class="n"><a id="t273" href="#t273">273</a></span><span class="t">                <span class="key">if</span> <span class="nam">e</span><span class="op">[</span><span class="num">1</span><span class="op">]</span> <span class="op">+</span> <span class="num">1</span> <span class="key">in</span> <span class="nam">cpgs_within_read_set</span>&nbsp;</span><span class="r"></span></p>
    <p class="run run2"><span class="n"><a id="t274" href="#t274">274</a></span><span class="t">            <span class="op">]</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t275" href="#t275">275</a></span><span class="t">&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t276" href="#t276">276</a></span><span class="t">            <span class="com"># If no CpGs covered (after filtering for matches only), skip</span>&nbsp;</span><span class="r"></span></p>
    <p class="par run show_par"><span class="n"><a id="t277" href="#t277">277</a></span><span class="t">            <span class="key">if</span> <span class="key">not</span> <span class="nam">this_segment_cpgs</span><span class="op">:</span>&nbsp;</span><span class="r"><span class="annotate short">277&#x202F;&#x219B;&#x202F;278</span><span class="annotate long">line 277 didn't jump to line 278 because the condition on line 277 was never true</span></span></p>
    <p class="mis show_mis"><span class="n"><a id="t278" href="#t278">278</a></span><span class="t">                <span class="key">continue</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t279" href="#t279">279</a></span><span class="t">&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t280" href="#t280">280</a></span><span class="t">            <span class="com"># Ok we're on the same strand as the methylation (right?)</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t281" href="#t281">281</a></span><span class="t">            <span class="com"># Let's compare the possible CpGs in this interval to the reference and note status</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t282" href="#t282">282</a></span><span class="t">            <span class="com">#   A methylated C will be *unchanged* and read as C (pair G)</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t283" href="#t283">283</a></span><span class="t">            <span class="com">#   An unmethylated C will be *changed* and read as T (pair A)</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t284" href="#t284">284</a></span><span class="t">            <span class="key">for</span> <span class="nam">query_pos</span><span class="op">,</span> <span class="nam">ref_pos</span> <span class="key">in</span> <span class="nam">this_segment_cpgs</span><span class="op">:</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t285" href="#t285">285</a></span><span class="t">                <span class="nam">query_base</span> <span class="op">=</span> <span class="nam">aligned_segment</span><span class="op">.</span><span class="nam">query_sequence</span><span class="op">[</span><span class="nam">query_pos</span><span class="op">]</span>  <span class="com"># type: ignore</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t286" href="#t286">286</a></span><span class="t">                <span class="com"># query_base_raw = aligned_segment.get_forward_sequence()[query_pos] # raw off sequencer</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t287" href="#t287">287</a></span><span class="t">                <span class="com"># query_base_no_offset = aligned_segment.query_alignment_sequence[query_pos] # this needs to be offset by the soft clip</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t288" href="#t288">288</a></span><span class="t">&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t289" href="#t289">289</a></span><span class="t">                <span class="com"># Store the read # in our sparse array</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t290" href="#t290">290</a></span><span class="t">                <span class="nam">coo_row</span><span class="op">.</span><span class="nam">append</span><span class="op">(</span><span class="nam">read_number</span><span class="op">)</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t291" href="#t291">291</a></span><span class="t">&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t292" href="#t292">292</a></span><span class="t">                <span class="com"># Store the CpG site in our sparse array</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t293" href="#t293">293</a></span><span class="t">                <span class="com"># TODO: Object orient these inputs? -- lots of bad inheritence style here</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t294" href="#t294">294</a></span><span class="t">                <span class="nam">coo_col</span><span class="op">.</span><span class="nam">append</span><span class="op">(</span>&nbsp;</span><span class="r"></span></p>
    <p class="run run2"><span class="n"><a id="t295" href="#t295">295</a></span><span class="t">                    <span class="nam">genome_methylation_embedding</span><span class="op">.</span><span class="nam">genomic_position_to_embedding</span><span class="op">(</span>&nbsp;</span><span class="r"></span></p>
    <p class="run run2"><span class="n"><a id="t296" href="#t296">296</a></span><span class="t">                        <span class="nam">chrom</span><span class="op">,</span>&nbsp;</span><span class="r"></span></p>
    <p class="run run2"><span class="n"><a id="t297" href="#t297">297</a></span><span class="t">                        <span class="nam">ref_pos</span> <span class="op">+</span> <span class="num">1</span><span class="op">,</span>&nbsp;</span><span class="r"></span></p>
    <p class="run run2"><span class="n"><a id="t298" href="#t298">298</a></span><span class="t">                    <span class="op">)</span>&nbsp;</span><span class="r"></span></p>
    <p class="run run2"><span class="n"><a id="t299" href="#t299">299</a></span><span class="t">                <span class="op">)</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t300" href="#t300">300</a></span><span class="t">&nbsp;</span><span class="r"></span></p>
    <p class="par run show_par"><span class="n"><a id="t301" href="#t301">301</a></span><span class="t">                <span class="key">if</span> <span class="nam">query_base</span> <span class="op">==</span> <span class="str">"C"</span><span class="op">:</span>&nbsp;</span><span class="r"><span class="annotate short">301&#x202F;&#x219B;&#x202F;303</span><span class="annotate long">line 301 didn't jump to line 303 because the condition on line 301 was never true</span></span></p>
    <p class="pln"><span class="n"><a id="t302" href="#t302">302</a></span><span class="t">                    <span class="com"># Methylated</span>&nbsp;</span><span class="r"></span></p>
    <p class="mis show_mis"><span class="n"><a id="t303" href="#t303">303</a></span><span class="t">                    <span class="nam">coo_data</span><span class="op">.</span><span class="nam">append</span><span class="op">(</span><span class="num">1</span><span class="op">)</span>&nbsp;</span><span class="r"></span></p>
    <p class="mis show_mis"><span class="n"><a id="t304" href="#t304">304</a></span><span class="t">                    <span class="key">if</span> <span class="nam">debug</span><span class="op">:</span>&nbsp;</span><span class="r"></span></p>
    <p class="mis show_mis"><span class="n"><a id="t305" href="#t305">305</a></span><span class="t">                        <span class="nam">print</span><span class="op">(</span><span class="fst">f"</span><span class="fst">\t</span><span class="op">{</span><span class="nam">query_pos</span><span class="op">}</span><span class="fst"> </span><span class="op">{</span><span class="nam">ref_pos</span><span class="op">}</span><span class="fst"> C-></span><span class="op">{</span><span class="nam">query_base</span><span class="op">}</span><span class="fst"> [Methylated]</span><span class="fst">"</span><span class="op">)</span>&nbsp;</span><span class="r"></span></p>
    <p class="par run show_par"><span class="n"><a id="t306" href="#t306">306</a></span><span class="t">                <span class="key">elif</span> <span class="nam">query_base</span> <span class="op">==</span> <span class="str">"T"</span><span class="op">:</span>&nbsp;</span><span class="r"><span class="annotate short">306&#x202F;&#x219B;&#x202F;307</span><span class="annotate long">line 306 didn't jump to line 307 because the condition on line 306 was never true</span></span></p>
    <p class="mis show_mis"><span class="n"><a id="t307" href="#t307">307</a></span><span class="t">                    <span class="nam">coo_data</span><span class="op">.</span><span class="nam">append</span><span class="op">(</span><span class="num">0</span><span class="op">)</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t308" href="#t308">308</a></span><span class="t">                    <span class="com"># Unmethylated</span>&nbsp;</span><span class="r"></span></p>
    <p class="mis show_mis"><span class="n"><a id="t309" href="#t309">309</a></span><span class="t">                    <span class="key">if</span> <span class="nam">debug</span><span class="op">:</span>&nbsp;</span><span class="r"></span></p>
    <p class="mis show_mis"><span class="n"><a id="t310" href="#t310">310</a></span><span class="t">                        <span class="nam">print</span><span class="op">(</span><span class="fst">f"</span><span class="fst">\t</span><span class="op">{</span><span class="nam">query_pos</span><span class="op">}</span><span class="fst"> </span><span class="op">{</span><span class="nam">ref_pos</span><span class="op">}</span><span class="fst"> C-></span><span class="op">{</span><span class="nam">query_base</span><span class="op">}</span><span class="fst"> [Unmethylated]</span><span class="fst">"</span><span class="op">)</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t311" href="#t311">311</a></span><span class="t">                <span class="key">else</span><span class="op">:</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t312" href="#t312">312</a></span><span class="t">                    <span class="nam">coo_data</span><span class="op">.</span><span class="nam">append</span><span class="op">(</span><span class="op">-</span><span class="num">1</span><span class="op">)</span>  <span class="com"># or just 0?</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t313" href="#t313">313</a></span><span class="t">                    <span class="key">if</span> <span class="nam">debug</span><span class="op">:</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t314" href="#t314">314</a></span><span class="t">                        <span class="nam">print</span><span class="op">(</span>&nbsp;</span><span class="r"></span></p>
    <p class="run run2"><span class="n"><a id="t315" href="#t315">315</a></span><span class="t">                            <span class="fst">f"</span><span class="fst">\t</span><span class="op">{</span><span class="nam">query_pos</span><span class="op">}</span><span class="fst"> </span><span class="op">{</span><span class="nam">ref_pos</span><span class="op">}</span><span class="fst"> C-></span><span class="op">{</span><span class="nam">query_base</span><span class="op">}</span><span class="fst"> [Unknown! SNV? Indel?]</span><span class="fst">"</span>&nbsp;</span><span class="r"></span></p>
    <p class="run run2"><span class="n"><a id="t316" href="#t316">316</a></span><span class="t">                        <span class="op">)</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t317" href="#t317">317</a></span><span class="t">&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t318" href="#t318">318</a></span><span class="t">            <span class="nam">read_number</span> <span class="op">+=</span> <span class="num">1</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t319" href="#t319">319</a></span><span class="t">&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t320" href="#t320">320</a></span><span class="t">            <span class="key">if</span> <span class="nam">debug</span><span class="op">:</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t321" href="#t321">321</a></span><span class="t">                <span class="nam">print</span><span class="op">(</span><span class="str">"************************************************\n"</span><span class="op">)</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t322" href="#t322">322</a></span><span class="t">&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t323" href="#t323">323</a></span><span class="t">                <span class="com"># query_bp = aligned_segment.query_sequence[pileupread.query_position]</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t324" href="#t324">324</a></span><span class="t">                <span class="com"># reference_bp = aligned_segment.get_reference_sequence()[aligned_segment.reference_start - pileupcolumn.reference_pos].upper()</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t325" href="#t325">325</a></span><span class="t">&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t326" href="#t326">326</a></span><span class="t">    <span class="com">## IIRC there's still a critical edge here, where sometimes we raise ValueError('row index exceeds matrix dimensions')</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t327" href="#t327">327</a></span><span class="t">&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t328" href="#t328">328</a></span><span class="t">    <span class="key">if</span> <span class="nam">debug</span><span class="op">:</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t329" href="#t329">329</a></span><span class="t">        <span class="nam">print</span><span class="op">(</span><span class="str">"Debug info for coo_matrix dimensions:"</span><span class="op">)</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t330" href="#t330">330</a></span><span class="t">        <span class="nam">print</span><span class="op">(</span><span class="fst">f"</span><span class="fst">\tcoo_row: </span><span class="op">{</span><span class="nam">len</span><span class="op">(</span><span class="nam">coo_row</span><span class="op">)</span><span class="op">:</span><span class="fst">,</span><span class="op">}</span><span class="fst">"</span><span class="op">)</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t331" href="#t331">331</a></span><span class="t">        <span class="nam">print</span><span class="op">(</span><span class="fst">f"</span><span class="fst">\tcoo row max: </span><span class="op">{</span><span class="nam">max</span><span class="op">(</span><span class="nam">coo_row</span><span class="op">)</span><span class="op">:</span><span class="fst">,</span><span class="op">}</span><span class="fst">"</span><span class="op">)</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t332" href="#t332">332</a></span><span class="t">        <span class="nam">print</span><span class="op">(</span><span class="fst">f"</span><span class="fst">\tcoo_col: </span><span class="op">{</span><span class="nam">len</span><span class="op">(</span><span class="nam">coo_col</span><span class="op">)</span><span class="op">:</span><span class="fst">,</span><span class="op">}</span><span class="fst">"</span><span class="op">)</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t333" href="#t333">333</a></span><span class="t">        <span class="nam">print</span><span class="op">(</span><span class="fst">f"</span><span class="fst">\tcoo col max: </span><span class="op">{</span><span class="nam">max</span><span class="op">(</span><span class="nam">coo_col</span><span class="op">)</span><span class="op">:</span><span class="fst">,</span><span class="op">}</span><span class="fst">"</span><span class="op">)</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t334" href="#t334">334</a></span><span class="t">        <span class="nam">print</span><span class="op">(</span><span class="fst">f"</span><span class="fst">\tcoo_data: </span><span class="op">{</span><span class="nam">len</span><span class="op">(</span><span class="nam">coo_data</span><span class="op">)</span><span class="op">:</span><span class="fst">,</span><span class="op">}</span><span class="fst">"</span><span class="op">)</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t335" href="#t335">335</a></span><span class="t">        <span class="nam">print</span><span class="op">(</span><span class="fst">f"</span><span class="fst">\tlen(read_name_to_row_number): </span><span class="op">{</span><span class="nam">len</span><span class="op">(</span><span class="nam">debug_read_name_to_row_number</span><span class="op">)</span><span class="op">:</span><span class="fst">,</span><span class="op">}</span><span class="fst">"</span><span class="op">)</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t336" href="#t336">336</a></span><span class="t">        <span class="nam">print</span><span class="op">(</span><span class="fst">f"</span><span class="fst">\ttotal_cpg_sites: </span><span class="op">{</span><span class="nam">genome_methylation_embedding</span><span class="op">.</span><span class="nam">total_cpg_sites</span><span class="op">:</span><span class="fst">,</span><span class="op">}</span><span class="fst">"</span><span class="op">)</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t337" href="#t337">337</a></span><span class="t">&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t338" href="#t338">338</a></span><span class="t">    <span class="com"># Generate the sparse matrix</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t339" href="#t339">339</a></span><span class="t">    <span class="nam">sparse_matrix</span> <span class="op">=</span> <span class="nam">scipy</span><span class="op">.</span><span class="nam">sparse</span><span class="op">.</span><span class="nam">coo_matrix</span><span class="op">(</span>&nbsp;</span><span class="r"></span></p>
    <p class="run run2"><span class="n"><a id="t340" href="#t340">340</a></span><span class="t">        <span class="op">(</span><span class="nam">coo_data</span><span class="op">,</span> <span class="op">(</span><span class="nam">coo_row</span><span class="op">,</span> <span class="nam">coo_col</span><span class="op">)</span><span class="op">)</span><span class="op">,</span>&nbsp;</span><span class="r"></span></p>
    <p class="run run2"><span class="n"><a id="t341" href="#t341">341</a></span><span class="t">        <span class="nam">shape</span><span class="op">=</span><span class="op">(</span><span class="nam">read_number</span><span class="op">,</span> <span class="nam">genome_methylation_embedding</span><span class="op">.</span><span class="nam">total_cpg_sites</span><span class="op">)</span><span class="op">,</span>&nbsp;</span><span class="r"></span></p>
    <p class="run run2"><span class="n"><a id="t342" href="#t342">342</a></span><span class="t">    <span class="op">)</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t343" href="#t343">343</a></span><span class="t">&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t344" href="#t344">344</a></span><span class="t">    <span class="com"># Validate a dimension of the coo_matrix, which should be:</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t345" href="#t345">345</a></span><span class="t">    <span class="com">#   Number of rows = number of reads that pass our filters</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t346" href="#t346">346</a></span><span class="t">    <span class="com">#   Number of columns = number of CpG sites</span>&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t347" href="#t347">347</a></span><span class="t">    <span class="key">assert</span> <span class="nam">sparse_matrix</span><span class="op">.</span><span class="nam">shape</span><span class="op">[</span><span class="num">1</span><span class="op">]</span> <span class="op">==</span> <span class="nam">genome_methylation_embedding</span><span class="op">.</span><span class="nam">total_cpg_sites</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t348" href="#t348">348</a></span><span class="t">&nbsp;</span><span class="r"></span></p>
    <p class="run"><span class="n"><a id="t349" href="#t349">349</a></span><span class="t">    <span class="key">return</span> <span class="nam">sparse_matrix</span>&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t350" href="#t350">350</a></span><span class="t">&nbsp;</span><span class="r"></span></p>
    <p class="pln"><span class="n"><a id="t351" href="#t351">351</a></span><span class="t">    <span class="com"># return scipy.sparse.coo_matrix((coo_data, (coo_row, coo_col)), shape=(len(read_name_to_row_number) + 1, total_cpg_sites))</span>&nbsp;</span><span class="r"></span></p>
</main>
<footer>
    <div class="content">
        <p>
            <a class="nav" href="z_21b632de40bb3de8_embedding_py.html">&#xab; prev</a> &nbsp; &nbsp;
            <a class="nav" href="index.html">&Hat; index</a> &nbsp; &nbsp;
            <a class="nav" href="z_a44f0ac069e85531_test_duplication_py.html">&#xbb; next</a>
            &nbsp; &nbsp; &nbsp;
            <a class="nav" href="https://coverage.readthedocs.io/en/7.13.1">coverage.py v7.13.1</a>,
            created at 2026-02-25 22:04 +0000
        </p>
    </div>
</footer>
</body>
</html>
